UI improvements

ntmmfts opened the pull request:
Resolving conflicts and adding updated code.

ntmmfts commented:
I reworked the action buttons for the checkbox functions - less intrusive and more usable.  Also added sorting for checkbox, pause/resume, and viewed/unviewed columns.  Also added the unviewed count to the title, e.g. "Change Detection - (3)".  Please take a look and let me know if you want any more fixes or have any more suggestions.  thnx.

ntmmfts commented:
And of course there's a leftover comment ending in watch-overview.  going to fix/push shortly.

ntmmfts commented:
Corrected.  Ready for review.

dgtlmoon commented on changedetectionio/templates/watch-overview.html:
```diff
@@ -27,18 +27,18 @@
 						<li>
 							<a href="javascript:processChecked('mark_selected_notviewed', '{{ active_tag }}');" class="pure-button button-tag " title="Mark Selected as Unviewed{%if active_tag%} in &quot;{{active_tag}}&quot;{%endif%}">Unviewed</a>
-						</li>
+						</li><br>
```
Br's dont belong in <UL>

ntmmfts replied:
Removed.  I added them for IE because I couldn't get display:-ms-grid to work, but I have it working now.

ntmmfts replied:
Noting that the fixed position grid display of the checkbox functions needs IE11 if IE is used.

dgtlmoon commented on changedetectionio/templates/watch-overview.html:
```diff
@@ -27,18 +27,18 @@
- -->			<div id="controls-top">
+			<div id="controls-top">
 				<ul id="post-list-buttons-top">
 					<span id="checkbox-functions" style="display: none;">
```
span dont belong in <UL>

ntmmfts replied:
Removed by reworking layout with divs and adjusting css.

dgtlmoon commented:
according to https://validator.w3.org/nu/#textarea there are a lot of defects here

![image](https://user-images.githubusercontent.com/275001/147382517-58acbe5d-1498-47cb-989d-92dad6cbfe1e.png)

maybe better is to use a small vuejs component.. i'm not sure about the overall quality here

dgtlmoon commented:
also please update your branch from the current `master`

ntmmfts commented:
Did lots of code cleanup and validation, so I'd like to avoid trying to figure out vuesjs for this effort.

dgtlmoon commented:
Alright thanks!

- Any chance of a `alt+s` shortcut to open the searchbox and focus on the input? that would be super helpful
- RSS icon spacing is not the same (Just remove the top one, it doesnt need to be there, its not in the current master) 
![image](https://user-images.githubusercontent.com/275001/147462119-1894aee9-249d-4937-bbf0-ede8cd2e90a8.png)
- Please remove the center-align of the text on the table

I'm thinking to merge this to my branch then I'll add some more fixes/ideas I have

dgtlmoon commented:
Some more..

### Use SVG
`static/images/search.png` please replace with a SVG

### Pause icon
![image](https://user-images.githubusercontent.com/275001/147471135-39a8abf9-80e6-4fa9-ac70-b1618dfdba42.png)
when it is NOT paused, please use a grey-icon, when I looked at it, it sort of felt like everything was already paused

### Read/unread
Less columns is always better, remove the read/unread, it takes up a lot of real-estate but it something thats not used much 

![image](https://user-images.githubusercontent.com/275001/147471513-c4519710-e1d5-43b3-945c-0eb6bd4705c3.png)

 migrate the icons to work in the title list and keep the bold/unbold technique


### Selector column can be joined with the count column

![image](https://user-images.githubusercontent.com/275001/147471720-d1941107-11ad-4550-8dd4-fff702d73716.png)

dgtlmoon commented:
Cutting down on those columns is going to help a lot when dealing with this little train wreck..

![image](https://user-images.githubusercontent.com/275001/147472260-b010ae45-f044-4cb1-8ffa-b0eabfca331b.png)

dgtlmoon commented:
Still seeing many validation issues

please paste into https://validator.w3.org/nu/#textarea
for example

![image](https://user-images.githubusercontent.com/275001/147473788-de9d82b8-763b-49c5-a2d1-ae21201322cd.png)

![image](https://user-images.githubusercontent.com/275001/147473822-85c628cb-68b9-4584-996d-5142ba716b4e.png)

ntmmfts commented:
working on fixes, should be done today.  "this little train wreck" comment, 🤣

dgtlmoon commented on changedetectionio/static/images/notviewed.svg:
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 25.4.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
+<svg version="1.1"
+	 id="svg2" xmlns:cc="http://creativecommons.org/ns#" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape" xmlns:ns1="http://sozi.baierouge.fr" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd" xmlns:svg="http://www.w3.org/2000/svg"
```
sozi "Sozi is a zooming presentation editor and player." :) :)

ntmmfts commented:
I think these fixes/changes address all of your latest comments:

- fixed all .svg's by embedding during save
- added ctl-alt-s hotkey to search (must be ctl- as browsers own alt-s and ctl-s
- added esc key clear search and focus to watch table
- removed text-align center in header/body/footer
- removed top recheck/rss buttons
- removed viewed/unviewed column
- merged # and checkbox columns
- reworked sort/search restriping and related css
- fixed validation errors
- fixed javascript formatting
- removed body onload event and added DOMContentLoaded event
- removed requests.args.get and get method in process_selected
	GET was a leftover from previous plan to call process_selected from mark_all_viewed 
	and index/pause_uiid to consolidate the code. 
- reworked checkbox-functions grid popup to be less intrusive and ensure visibility in desktop and mobile
- reworked pause_uuid code

Re: "Why session storage and not a cookie here? whats the benefits, just out of curiosity?"

	I wanted to make the sort column & order was temporary to make sure I don't miss any unviewed watches.
	Using local storage makes the last_changed / descending sorting return if the browser or tab is closed.

dgtlmoon commented:
other thought i had was, you should test this with like 5,000 entries, just to be sure it doesnt run slow/give some unexpected error

ntmmfts commented:
Sorry about that, I had forgotten to rerun tests.  Strange issue, the code I added to correct flash message grammar 
fails in the otherwise-unaltered api_watch_checknow route despite working fine in the process_checked route.  I tried multiple workarounds with different format constructs including f-Strings, but couldn't make it pass even though they work during usage - not sure why, but it seems like the test queue is emptying down to 1 watch and it shouldn't.  If you could check to see if it can be fixed to retain the grammar formatting I'd appreciate it.  thnx!

ntmmfts commented:
Not sure what's happening, but now b129290 fails with "assert b'Test Body Value' in b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD H..."  

Might be because I started using worktrees for the two separate efforts, ui-improvements and seconds, but both worktrees are created from an up to date master and started with the latest working commits.  Maybe one of the master commits since I started working in the (detached?) ui-improvments branch broke it.  Haven't figure out what's actually causing the failure yet, so I might have to go back to a non-worktree environment with the last working commit, b129290, unless you know what exactly might be causing the failure?

Or, maybe you can take it from the last good commit and work whatever ideas/changes you want so it can finally get into master the way you want it?

Thnx.

dgtlmoon commented:
@ntmmfts i see in the logs, that they both fail in the same place, there is no mystery :) visiting the `[PREVIEW]` link is broken, the test uses `/preview/first` just to make its life easier (cause it doesnt know the UUID)

```
        # The service should echo back the body
        res = client.get(
            url_for("preview_page", uuid="first"),
            follow_redirects=True
        )
    
        # Check if body returned contains the specified data
>       assert str.encode(body_value) in res.data
E       assert b'Test Body Value' in b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>500 Internal Server Error</title>\n<h1>Internal Serv...nd was unable to complete your request. Either the server is overloaded or there is an error in the application.</p>\n'
E        +  where b'Test Body Value' = <method 'encode' of 'str' objects>('Test Body Value')
E        +    where <method 'encode' of 'str' objects> = str.encode
E        +  and   b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>500 Internal Server Error</title>\n<h1>Internal Serv...nd was unable to complete your request. Either the server is overloaded or there is an error in the application.</p>\n' = <WrapperTestResponse 290 bytes [500 INTERNAL SERVER ERROR]>.data
```

```
tests/test_request.py::test_body_in_request 
-------------------------------- live log call ---------------------------------
18:35:32 root: INFO Saving JSON..
18:35:32 root: INFO Saving JSON..
18:35:33 root: INFO Saving JSON..
18:35:33 urllib3.connectionpool: DEBUG Starting new HTTP connection (1): localhost:5005
18:35:33 urllib3.connectionpool: DEBUG Starting new HTTP connection (1): localhost:5005
18:35:33 urllib3.connectionpool: DEBUG Starting new HTTP connection (1): localhost:5005
18:35:33 werkzeug: INFO 127.0.0.1 - - [05/Jan/2022 18:35:33] "GET /test-body HTTP/1.1" 200 -
18:35:33 werkzeug: INFO 127.0.0.1 - - [05/Jan/2022 18:35:33] "GET /test-body HTTP/1.1" 200 -
18:35:33 urllib3.connectionpool: DEBUG http://localhost:5005 "GET /test-body HTTP/1.1" 200 0
18:35:33 werkzeug: INFO 127.0.0.1 - - [05/Jan/2022 18:35:33] "GET /test-body HTTP/1.1" 200 -
18:35:33 urllib3.connectionpool: DEBUG http://localhost:5005 "GET /test-body HTTP/1.1" 200 0
18:35:33 urllib3.connectionpool: DEBUG http://localhost:5005 "GET /test-body HTTP/1.1" 200 0
18:35:35 urllib3.connectionpool: DEBUG Starting new HTTP connection (1): localhost:5005
18:35:35 urllib3.connectionpool: DEBUG Starting new HTTP connection (1): localhost:5005
18:35:35 werkzeug: INFO 127.0.0.1 - - [05/Jan/2022 18:35:35] "GET /test-body HTTP/1.1" 200 -
18:35:35 werkzeug: INFO 127.0.0.1 - - [05/Jan/2022 18:35:35] "GET /test-body HTTP/1.1" 200 -
18:35:35 urllib3.connectionpool: DEBUG http://localhost:5005 "GET /test-body HTTP/1.1" 200 0
18:35:35 urllib3.connectionpool: DEBUG http://localhost:5005 "GET /test-body HTTP/1.1" 200 0
18:35:38 changedetectionio: ERROR Exception on /preview/first [GET]
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/site-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/site-packages/flask_login/utils.py", line 269, in decorated_view
    return func(*args, **kwargs)
  File "/home/runner/work/changedetection.io/changedetection.io/changedetectionio/__init__.py", line 839, in preview_page
    newest = list(watch['history'].keys())[-1]
IndexError: list index out of range
FAILED                   
```

dgtlmoon commented:
@ntmmfts how are you feeling about the changes so far? I was thinking it should have some selenium test to check the stuff works as expected. I have some small changes todo with font-sizing and padding.. so I was thinking to merge this into `ui-improvements` and then add my own small changes and tests ontop

dgtlmoon commented:
grrrrr i pushed your branch into the `ui-improvements` branch so we can both work on our own branches but it closed the PR, can you add another commit and we'll see if it will re-open the PR?
